File: Data Structures & Algorithms/is-palindrome/submission-2.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isPalindrome(s) {

        function isAlphaNum(char) {
            return 'a' <= char && char <= 'z' || 'A' <= char && char <= 'Z' || '0' <= char && char <= '9'; 
        }
        let left = 0;
        let right = s.length -1;

        while (left < right) {
            if (isAlphaNum(s[left]) && isAlphaNum(s[right])) {
                if (s[left].toLowerCase() === s[right].toLowerCase()) {
                    left++;
                    right--;
                    continue;
                }
                else return false;
            }
            else if (isAlphaNum(s[left])) {
                right--;
            }
            else left++;
        }
        return true;
            }
}

File: Data Structures & Algorithms/sliding-window-maximum/submission-2.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        """
         maximum element in the window
         lazy deetion
         1 heaps
         heap  -> biggest elem -> + idx
         (1,0)
         (2,1)
        """
        delete = set()
        n = len(nums)
        heap = []
        left = 0
        ans = []
        for right, num in enumerate(nums):
            heapq.heappush(heap, (-num, right))

            window_length = right - left + 1
            if window_length < k:
                continue
            if window_length > k:
                delete.add((-nums[left], left))
                left+=1
                
            while heap and heap[0] in delete:
                heapq.heappop(heap)
            ans.append(-heap[0][0])
        return ans


